timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeLimitedLockService.java

```diff
@@ -0,0 +1,224 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.timelock.lock;
+
+import java.math.BigInteger;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+import javax.ws.rs.PathParam;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.SimpleTimeLimiter;
+import com.google.common.util.concurrent.TimeLimiter;
+import com.google.common.util.concurrent.UncheckedTimeoutException;
+import com.palantir.lock.HeldLocksGrant;
+import com.palantir.lock.HeldLocksToken;
+import com.palantir.lock.LockClient;
+import com.palantir.lock.LockRefreshToken;
+import com.palantir.lock.LockRequest;
+import com.palantir.lock.LockResponse;
+import com.palantir.lock.LockServerOptions;
+import com.palantir.lock.LockService;
+import com.palantir.lock.SimpleHeldLocksToken;
+import com.palantir.lock.remoting.BlockingTimeoutException;
+
+public class BlockingTimeLimitedLockService implements LockService {
+    private static final Logger log = LoggerFactory.getLogger(BlockingTimeLimitedLockService.class);
+
+    private final LockService delegate;
+    private final TimeLimiter timeLimiter;
+    private final long blockingTimeLimitMillis;
+
+    @VisibleForTesting
+    BlockingTimeLimitedLockService(
+            LockService delegate,
+            TimeLimiter timeLimiter,
+            long blockingTimeLimitMillis) {
+        this.delegate = delegate;
+        this.timeLimiter = timeLimiter;
+        this.blockingTimeLimitMillis = blockingTimeLimitMillis;
+    }
+
+    public static LockService create(LockService lockService, long blockingTimeLimitMillis) {
+        return new BlockingTimeLimitedLockService(lockService, new SimpleTimeLimiter(), blockingTimeLimitMillis);
+    }
+
+    @Nullable
+    @Override
+    public LockRefreshToken lock(@PathParam("client") String client, LockRequest request) throws InterruptedException {
+        return callWithTimeLimit(
+                () -> delegate.lock(client, request),
+                ImmutableLockRequestSpecification.of("lock", client, request));
+    }
+
+    @Override
+    public HeldLocksToken lockAndGetHeldLocks(@PathParam("client") String client, LockRequest request)
+            throws InterruptedException {
+        return callWithTimeLimit(
+                () -> delegate.lockAndGetHeldLocks(client, request),
+                ImmutableLockRequestSpecification.of("lockAndGetHeldLocks", client, request));
+    }
+
+    @Override
+    public boolean unlock(LockRefreshToken token) {
```

Comment:
Hmm, it might be fairly high lift to do this actually, since the methods in question aren't interruptible (so even if you pass them through the time-limiter they will keep going)! We haven't seen this occur in production so I doubt it would be a problem.

My main concern is perf related (though might be a premature optimisation) - because firing off such a request actually spins up a separate thread to run the task (and interrupting the thread if it goes overtime). Happy to discuss this one offline.